compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -10513,16 +10510,73 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
             *returnCond = TR::InstOpCode::COND_MASK0;
          }
       }
+
    else if (isConstLength)
       {
+      TR::LabelSymbol * cFlowRegionEnd;
       if (node->getOpCode().hasSymbolReference() && node->getSymbolReference())
          {
          baseSource1Ref = generateS390MemoryReference(cg, node, source1Node, 0, true); // forSS=true
          baseSource2Ref = generateS390MemoryReference(cg, node, source2Node, 0, true); // forSS=true
+         if (baseSource1Ref->getOffset() == baseSource2Ref->getOffset())
+            {
+            // since forSS was true in the generation of the memory references, there will not be any index registers.
+            // if there are base registers, we can compare those and conclude if are equal.
+            if(baseSource1Ref->getBaseRegister() && baseSource2Ref->getBaseRegister())
+               {
+               generateS390CompareAndBranchInstruction(
+                  cg,
+                  TR::InstOpCode::getCmpLogicalRegOpCode(),
+                  node,
+                  baseSource1Ref->getBaseRegister(),
+                  baseSource2Ref->getBaseRegister(),
+                  TR::InstOpCode::COND_BE,
+                  isIfxcmpBrCondContainEqual ? compareTarget : cFlowRegionEnd);
+               }
+            // if there is no base registerand offsets are equal, we can return.
+            else if((NULL == baseSource1Ref->getBaseRegister()) && (NULL == baseSource2Ref->getBaseRegister()))
+               {
+               if (isIfxcmpBrCondContainEqual)
+                  {
+                  generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_MASK15, node, compareTarget, deps);
+                  }
+
+               cg->recursivelyDecReferenceCount(lengthNode);
+               cg->processUnusedNodeDuringEvaluation(source1Node);
+               cg->processUnusedNodeDuringEvaluation(source2Node);
+
+               if (returnCond!=NULL)
+                  {
+                  if (isEqualCmp)
+                     *returnCond = TR::InstOpCode::COND_MASK15;
+                  else
+                     *returnCond = TR::InstOpCode::COND_MASK0;
+                  }
+
+               if (isFoldedIf)
+                  {
+                  cg->decReferenceCount(node);
+                  }
+               else if (needResultReg)
+                  {
+                  node->setRegister(retValReg);
+                  }
+               return retValReg;
+               }
+            }
          }
       else
          {
          // memcmp2 lacks a symbol reference
+         generateS390CompareAndBranchInstruction(
+            cg,
+            TR::InstOpCode::getCmpLogicalRegOpCode(),
+            node,
+            cg->evaluate(source1Node),
```

Comment:
we evaluate here as well as in the creation of mem ref. Maybe makes more sense assign it to a named register first. I can Also move this code after mem ref creation and compare base registers of mem refs!